src/Native/Unix/System.Native/pal_io.c

```diff
@@ -1265,6 +1285,23 @@ int32_t SystemNative_CopyFile(intptr_t sourceFd, const char* srcPath, const char
     fcntl(outFd, F_SETFD, FD_CLOEXEC);
 #endif
 
+    while ((ret = flock(outFd, LOCK_EX | LOCK_NB)) < 0 && errno == EINTR);
```

Comment:
For Linux, it might be a better idea to use O_TMPFILE to create a temporary file while the copy is being performed, and when it's done, hardlink it with the new name.  Then there's no need to use advisory file locking, which isn't really that useful.

Reply:
The advisory locking is used by the managed FileStream so whatever is used it has to be done on both sides and match.

Reply:
I would still like us to try hard to find ways around duplicating all this logic.